### DISCUSSION TITLE
After loading new page in current Tab the url_changed SIGNAL is not emited

### DISCUSSION THREAD

**Discussion by Stas1981ru** [qutebrowser/qutebrowser#5705] [Q&A] [answered]
Example:
cur_tab=objreg.window_by_index(0).tabbed_browser._now_focused - it is then current tab
cur_tab.url_changed.connect(...) - we connect SLOT A to "url_changed" SIGNAL of the tab

1 Load some url: ":open url"   - it is the first Page in hostory of the QWebEngineView
2 SLOT A is working good
3 Now we are click on some links on the current page (page of the QtWebEngineView is reloaded) without new tab open
4 SLOT A is NOT working now (load_started,load_finished still working good)

It is a bug, or my misunderstanding ?
I think, the SLOT A is still connected to the first Page in the hystory of the QWebEngineView after load a new page in current tab.

**Comment by The-Compiler** [comment, marked as answer]
What exactly are you trying to do? Do you have a more complete example? This hack in a config.py:

```python
from qutebrowser.utils import objreg
from qutebrowser.api import cmdutils

@cmdutils.register()
def setup_handler():
    cur_tab=objreg.window_by_index(0).tabbed_browser._now_focused
    cur_tab.url_changed.connect(lambda url: print(f"url changed to {url}"))
```

works fine for me.

**Reply by Stas1981ru** [reply]
>What exactly are you trying to do?
Manipulation in qb browser from other application like a real user. (qb transmit to main application a positions of DOC elements. Main application get it and click/move/scroll/... with real mouse pointer)

I put you hack in a config.py and get at start:
`...
20:49:07 WARNING: Function config.setup_handler() from /home/user/my/seo_instrument/python/2018_new/backup/2020chrom/config_qute/config.py has no docstring
...`

It print nothing for me while browsing )
I try to test it again with fresh install qb later.

**Reply by The-Compiler** [reply]
You can ignore the warning. Note you'll need to run the `:setup-handler` qutebrowser command for it to start printing things.

**Reply by Stas1981ru** [reply]
> You can ignore the warning. Note you'll need to run the `:setup-handler` qutebrowser command for it to start printing things.

Thanks, you are right all works fine. In is mistake from my side. It is a floating bug, I can't find it now.
Can I discuss the integration qb in this thread, so as not to clog with new themes?

**Reply by The-Compiler** [reply]
> Can I discuss the integration qb in this thread, so as not to clog with new themes?

 Go ahead, but it's probably best if you post a new reply below so we can have one thread for every question in this topic.

**Comment by Stas1981ru** [comment]
I have a question about rigth integration of qt to other application.

Now I make it with some hack:

1 in file ../qutebrowser/qutebrowser/app.py

```
def qt_mainloop():
     ...
    -return q_app.exec_()
    +return q_app
```


2 in my main python application:

```
import qutebrowser.qutebrowser
from PyQt5.QtCore import QEventLoop
from qutebrowser.utils import objreg
...
q_app=qutebrowser.qutebrowser.main()
cur_window=objreg.window_by_index(0)
cur_window._commandrunner.run(":open -t https://www.youtube.com/watch?v=h9uv7XgP8Io")
...
#small ugly example for tests only
i=1
while i<1950000:
	q_app.processEvents(QEventLoop.AllEvents,1000)
	i+=1
...
#closing the current (in example it is the last opened - youtube tab) tab
cur_window._commandrunner.run(":tab-close")

#time to see what heppened after command
i=1
while i<95000000:
	#print (i)    
	q_app.processEvents(QEventLoop.AllEvents,1000)
	i+=1
```

My question:
I find same ways to close the tab (they all are works fine) :
```
@cmdutils.register()
def test_close1():
	cur_window=objreg.window_by_index(0)
	cur_window.tabbed_browser.close_tab(cur_window.tabbed_browser._now_focused)

@cmdutils.register()
def test_close2():
	cur_window=objreg.window_by_index(0)
	cur_window._commandrunner.run(":tab-close")

@cmdutils.register()
def test_close3():
	cur_window=objreg.window_by_index(0)
	cur_window.tabbed_browser.widget.tabCloseRequested.emit(cur_window.tabbed_browser.widget.count() - 1)
```

Close by middle mouse button in q_app.processEvents(QEventLoop.AllEvents,1000) is working fine too.
All commands (:test_close1,:test_close2,:test_close3) are works fine from cammand prompt of qb too.


But ! Code in my example is not working well, the tab is closing on visually. Youtube video still working (We can hear the sound only) and we can see it "closed" tab in QTWEBENGINE_REMOTE_DEBUGGING view.

I need to close current tab from my thread, can you help me with that problem ?

**Reply by The-Compiler** [reply]
> I need to close current tab from my thread, can you help me with that problem ?

Doing Qt GUI operations from a non-main thread is undefined behavior.

**Reply by Stas1981ru** [reply]
> > I need to close current tab from my thread, can you help me with that problem ?
> 
> Doing Qt GUI operations from a non-main thread is undefined behavior.

But my thread is __main__ , I start qb here : q_app=qutebrowser.qutebrowser.main() from MY main thread.

**Reply by The-Compiler** [reply]
No idea then. Not much more I can say without a complete example I can actually run myself to reproduce the issue. But even then, this is totally unsupported. What's your goal exactly? You might want to use QtWebEngine directly (which gives you a stable API) rather than trying to embed qutebrowser.

**Reply by Stas1981ru** [reply]
>You might want to use QtWebEngine directly
I alrady use it, but now I try with full supported browser.

For reproduce my "closing" you must 
1 put file "close_test.py" to parent directory of qb (where setup.py is located)

```
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
from PyQt5.QtWidgets import QApplication
import qutebrowser.qutebrowser
from qutebrowser.utils import objreg
from PyQt5.QtCore import QEventLoop
import time

q_app=qutebrowser.qutebrowser.main()
cur_window=objreg.window_by_index(0)
#open new tab (it will be second tab of qb)
cur_window._commandrunner.run(":open -t https://www.youtube.com/watch?v=kRptDuN8NrY")

itime = time.time()
#10 seconds to load youtube
while time.time() - itime < 10:
	q_app.processEvents(QEventLoop.AllEvents,1000)

print ("Tab was closed")
cur_window._commandrunner.run(":tab-close")

#15 seconds to see reaction
while time.time() - itime < 15:
	q_app.processEvents(QEventLoop.AllEvents,1000)

cur_window._commandrunner.run(":quit")
```

2. in file ../qutebrowser/qutebrowser/app.py
```
def qt_mainloop():
     ...
    -return q_app.exec_()
    +return q_app
```
